man/man3/nnetcdf.3
.\" 
.\" Copyright (c) 2009 Gordon D. Carrie
.\" All rights reserved.
.\" 
.\" Please address questions and feedback to dev0@trekix.net
.\" 
.\" $Revision: 1.2 $ $Date: 2013/10/04 16:22:27 $
.\"
.TH nnetcdf 3 "NetCDF convenience functions"
.SH NAME
NNC_Open, NNC_Inq_Dim, NNC_Get_Var_Text, NNC_Get_String, NNC_Get_Var_Uchar, NNC_Get_Var_Int, NNC_Get_Var_UInt, NNC_Get_Var_Float, NNC_Get_Var_Double, NNC_Get_Att_String, NNC_Get_Att_Int, NNC_Get_Att_UInt, NNC_Get_Att_Float \- NetCDF convenience functions
.SH SYNOPSIS
.nf
\fB#include "nnetcdf.h"\fP
\fB#define NNCDF_ERROR 1\fP
\fBint\fP \fBNNC_Open\fP(\fBchar *\fP\fIfile_nm\fP, \fBjmp_buf\fP \fIerror_env\fP);
\fBsize_t\fP \fBNNC_Inq_Dim\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIdim_name\fP, \fBjmp_buf\fP \fIerror_env\fP);
\fBchar *\fP \fBNNC_Get_String\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBjmp_buf\fP \fIerror_env\fP);
\fBchar *\fP\fBNNC_Get_Var_Text\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBchar *\fP\fIcPtr\fP, \fBjmp_buf\fP \fIerror_env\fP);
\fBunsigned\fP \fBchar *\fP \fBNNC_Get_Var_UChar\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBunsigned\fP \fBchar *\fP\fIuPtr\fP,
    \fBjmp_buf\fP \fIerror_env\fP);
\fBint *\fP \fBNNC_Get_Var_Int\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBint *\fP\fIiPtr\fP, \fBjmp_buf\fP \fIerror_env\fP);
\fBunsigned *\fP \fBNNC_Get_Var_UInt\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBunsigned *\fP\fIiPtr\fP, \fBjmp_buf\fP \fIerror_env\fP);
\fBfloat *\fP \fBNNC_Get_Var_Float\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBfloat *\fP\fIfPtr\fP,
    \fBjmp_buf\fP \fIerror_env\fP);
\fBdouble *\fP \fBNNC_Get_Var_Double\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBdouble *\fP\fIdPtr\fP,
    \fBjmp_buf\fP \fIerror_env\fP);
\fBchar *\fP \fBNNC_Get_Att_String\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBchar *\fP\fIatt\fP,
    \fBjmp_buf\fP \fIerror_env\fP);
\fBint *\fP \fBNNC_Get_Att_Int\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBchar *\fP\fIatt\fP, \fBjmp_buf\fP \fIerror_env\fP);
\fBfloat *\fP \fBNNC_Get_Att_Float\fP(\fBint\fP \fIncid\fP, \fBchar *\fP\fIvar_name\fP, \fBchar *\fP\fIatt\fP, \fBjmp_buf\fP \fIerror_env\fP);
.fi
.SH DESCRIPTION
The following functions simplify access to some NetCDF functions.

If a function takes an \fIerror_env\fP argument, it should follow a call
to \fBsetjmp()\fP with \fIerror_env\fP as the \fBjmp_buf\fP argument.
If something goes wrong, the function will use \fBlongjmp()\fP to
repeat the \fBsetjmp()\fP call, but with a return value of \fBNNCDF_ERROR\fP.
The program should then go to a suitable error branch.  See the
documentation for \fBsetjmp\fP and \fBlongjmp\fP for more information.

\fBNNC_Open()\fP opens a NetCDF file named \fIfile_nm\fP and returns a NetCDF
file identifier for it.
It uses \fIerror_env\fP to handle errors as described above. It prints error
messages to \fBstderr\fP.
The file should eventually be closed with a call to \fBnc_close()\fP.

\fBNNC_Inq_Dim()\fP returns the size of the dimension named \fIdim_name\fP
from the NetCDF file identified as \fBncid\fP, which should be a return value
from \fBNNC_Open()\fP or \fBnc_open()\fP.
It uses \fIerror_env\fP to handle errors as described above.  It prints error
messages to \fBstderr\fP.

\fBNNC_Get_String()\fP returns a nul terminated string corresponding to variable
named \fIvar_name\fP from the NetCDF file identified as \fBncid\fP, which should
be a return value from \fBNNC_Open()\fP or \fBnc_open()\fP.
Memory is allocated for the returned string.  This memory should eventually
be freed with a call to \fBFREE()\fP.
It uses \fIerror_env\fP to handle errors as described above.  It prints error
messages to \fBstderr\fP.

\fBNNC_Get_Var_Text()\fP returns the character array corresponding to variable
\fIvar_name\fP from the NetCDF file identified as \fBncid\fP, which should be a
return value from \fBNNC_Open()\fP or \fBnc_open()\fP.
If \fIcPtr\fP is \fBNULL\fP, memory is allocated for the returned array.  This
memory should eventually be freed with a call to \fBFREE()\fP.  If \fIcPtr\fP
is not \fBNULL\fP, it should be a pointer to sufficient storage for the array.
Then \fIcPtr\fP is returned with the values copied into it.
It uses \fIerror_env\fP to handle errors as described above.  It prints error
messages to \fBstderr\fP.

\fBNNC_Get_Var_UChar()\fP is like \fBNNC_Get_Var_Text()\fP, except that it
retrieves an array of unsigned characters.

\fBNNC_Get_Var_Int()\fP is like \fBNNC_Get_Var_Text()\fP, except that it
retrieves an array of integers.

\fBNNC_Get_Var_UInt()\fP is like \fBNNC_Get_Var_Text()\fP, except that it
retrieves an array of unsigned integers.

\fBNNC_Get_Var_Float()\fP is like \fBNNC_Get_Var_Text()\fP, except that it
retrieves an array of floats.

\fBNNC_Get_Var_Double()\fP is like \fBNNC_Get_Var_Text()\fP, except that it
retrieves an array of doubles.

\fBNNC_Get_Att_String()\fP returns a nul terminated string attribute for the
variable named \fIvar_name\fP in the NetCDF file identified as \fBncid\fP, which
should be a return value from \fBNNC_Open()\fP or \fBnc_open()\fP.
If \fIvar_name\fP is "\fBNC_GLOBAL\fP", a global attribute is sought.
Memory is allocated for the returned string.  This memory should eventually
be freed with a call to \fBFREE()\fP.
It uses \fIerror_env\fP to handle errors as described above.  It prints error
messages to \fBstderr\fP.

\fBNNC_Get_Att_Int()\fP returns an integer attribute for the variable named
\fIvar_name\fP in the NetCDF file identified as \fBncid\fP, which should be a
return value from \fBNNC_Open()\fP or \fBnc_open()\fP.
If \fIvar_name\fP is "\fBNC_GLOBAL\fP", a global attribute is sought.
Memory is allocated for the returned array.  This memory should eventually
be freed with a call to \fBFREE()\fP.
It uses \fIerror_env\fP to handle errors as described above.  It prints error
messages to \fBstderr\fP.

\fBNNC_Get_Att_UInt()\fP returns an unsigned integer attribute for the variable
named \fIvar_name\fP in the NetCDF file identified as \fBncid\fP, which should
be a return value from \fBNNC_Open()\fP or \fBnc_open()\fP.
If \fIvar_name\fP is "\fBNC_GLOBAL\fP", a global attribute is sought.
Memory is allocated for the returned array.  This memory should eventually
be freed with a call to \fBFREE()\fP.
It uses \fIerror_env\fP to handle errors as described above.  It prints error
messages to \fBstderr\fP.

\fBNNC_Get_Att_Float()\fP is like \fBNNC_Get_Att_Int()\fP, except that it
returns a float attribute.
.SH SEE ALSO
\fBnetcdf\fP (3), \fBsetjmp\fP (3), \fBlongjmp\fP (3), \fBalloc\fP (3)
.SH AUTHOR
Copyright (c) 2007, 2008, 2009 Gordon D. Carrie.  All rights reserved.


RELEASE_NOTES
Pre-MultiDop DDA Release Notes

20150616014655
Initial check in. Previous history is in RCS.tar.gz

20160219162111
DDA.h --
    Incremented release version to 0.8.2.
--
DDA.c --
    Parameter file can now provide a name for the velocity variable in OPAWS
    input. If none given, DDA defaults to "VT".
--
geog_proj.c geog_proj.3 --
    GeogProjLonLatToXY and GeogProjXYToLonLat set return arguments to NAN on
    failure.
--
geog_lib.c geog_lib.3 --
    Set init variable when setting Earth radius from GEOG_REARTH.
    In other words, set Earth radius from environment variable once only.
--

20160516171935
unix_defs.h --
    Set C99_REQUIRED compiler variable to force c99, otherwise, c99 is optional.
--



src/nnetcdf.c
/*
   -	nnetcdf.c --
   -		This file defines functions that read
   -		NetCDF files.  See nnetcdf (3).
   -	
   .	Copyright (c) 2011, Gordon D. Carrie. All rights reserved.
   .	
   .	Redistribution and use in source and binary forms, with or without
   .	modification, are permitted provided that the following conditions
   .	are met:
   .	
   .	    * Redistributions of source code must retain the above copyright
   .	    notice, this list of conditions and the following disclaimer.
   .
   .	    * Redistributions in binary form must reproduce the above copyright
   .	    notice, this list of conditions and the following disclaimer in the
   .	    documentation and/or other materials provided with the distribution.
   .	
   .	THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   .	"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   .	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   .	A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
   .	HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   .	SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
   .	TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   .	PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   .	LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   .	NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   .	SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
   .
   .	Please send feedback to dev0@trekix.net
 */

#include <string.h>
#include <stdio.h>
#include "alloc.h"
#include "nnetcdf.h"

/* Open a NetCDF file. See nnetcdf (3). */
int NNC_Open(const char *file_nm, jmp_buf error_env)
{
    int status;
    int ncid;

    if ((status = nc_open(file_nm, 0, &ncid)) != 0) {
	fprintf(stderr, "Could not open %s. NetCDF error message is: %s\n",
		file_nm ? file_nm : "(NULL)", nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return ncid;
}

/* Return the size of a NetCDF dimension.  See nnetcdf (3). */
size_t NNC_Inq_Dim(int ncid, const char *name, jmp_buf error_env)
{
    int dimid;
    int status;
    size_t len;

    if ((status = nc_inq_dimid(ncid, name, &dimid)) != 0) {
	fprintf(stderr, "Could not find dimension named %s. "
		"NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_inq_dim(ncid, dimid, NULL, &len)) != 0) {
	fprintf(stderr, "Could not retrieve size of %s dimension.  "
		"NetCDF error message is: %s\n", name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return len;
}

/* Return a string from a NetCDF file. See nnetcdf (3). */
char * NNC_Get_String(int ncid, const char *name, jmp_buf error_env)
{
    char *val;		/* Return value */
    int varid;		/* Variable identifier */
    int dimid;		/* Dimension of variable */
    size_t len;		/* Dimension size */
    char *c, *ce;	/* Loop parameters */
    int status;		/* NetCDF function return value */

    if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_inq_vardimid(ncid, varid, &dimid)) != 0) {
	fprintf(stderr, "Could not get dimension for %s. "
		"NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_inq_dimlen(ncid, dimid, &len)) != 0) {
	fprintf(stderr, "Could not get dimension length for %s. "
		"NetCDF error message is: %s\n", name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !(val = MALLOC(len + 1)) ) {
	fprintf(stderr, "Allocation failed for %s.\n", name);
	longjmp(error_env, NNCDF_ERROR);
    }
    for (c = val, ce = c + len; c < ce; c++) {
	*c = ' ';
    }
    if ((status = nc_get_var_text(ncid, varid, val)) != 0) {
	fprintf(stderr, "Could not get value for %s. NetCDF error message "
		"is: %s\n", name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    *(val + len) = '\0';

    return val;
}

/* Retrieve a character variable from a NetCDF file. See nnetcdf (3). */
char *NNC_Get_Var_Text(int ncid, const char *name, char *cPtr,
	jmp_buf error_env)
{
    int varid;		/* Variable identifier */
    int status;

    if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !cPtr ) {
	int ndims;
	static int *dimidPtr;
	int *d, *de;
	size_t sz;

	if ((status = nc_inq_varndims(ncid, varid, &ndims)) != 0) {
	    fprintf(stderr, "Could not get dimension count for %s. "
		    "NetCDF error message is: %s\n",
		    name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ( !(dimidPtr = REALLOC(dimidPtr, ndims * sizeof(int))) ) {
	    fprintf(stderr, "Could not allocate %d dimension array for %s\n",
		    ndims, name);
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ((status = nc_inq_vardimid(ncid, varid, dimidPtr)) != 0) {
	    fprintf(stderr, "Could not get dimensions for %s. "
		    "NetCDF error message is: %s\n", name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	for (sz = 1, d = dimidPtr, de = d + ndims; d < de; d++) {
	    size_t l;

	    if ((status = nc_inq_dimlen(ncid, *d, &l)) != 0) {
		fprintf(stderr, "Could not get dimension size for %s. "
			"NetCDF error message is: %s\n",
			name, nc_strerror(status));
		longjmp(error_env, NNCDF_ERROR);
	    }
	    sz *= l;
	}
	if ( !(cPtr = MALLOC(sz)) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
    }
    if ((status = nc_get_var_text(ncid, varid, cPtr)) != 0) {
	fprintf(stderr, "Could not get value for %s. "
		"NetCDF error message is: %s\n", name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return cPtr;
}

/* Retrieve an unsigned char variable from a NetCDF file. See nnetcdf (3). */
unsigned char * NNC_Get_Var_UChar(int ncid, const char *name,
	unsigned char *uPtr, jmp_buf error_env)
{
    int varid;		/* Variable identifier */
    int status;

    if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !uPtr ) {
	int ndims;
	static int *dimidPtr;
	int *d, *de;
	size_t sz;

	if ((status = nc_inq_varndims(ncid, varid, &ndims)) != 0) {
	    fprintf(stderr, "Could not get dimension count for %s. "
		    "NetCDF error message is: %s\n",
		    name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ( !(dimidPtr = REALLOC(dimidPtr, ndims * sizeof(int))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ((status = nc_inq_vardimid(ncid, varid, dimidPtr)) != 0) {
	    fprintf(stderr, "Could not get dimensions for %s. "
		    "NetCDF error message is: %s\n", name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	for (sz = 1, d = dimidPtr, de = d + ndims; d < de; d++) {
	    size_t l;

	    if ((status = nc_inq_dimlen(ncid, *d, &l)) != 0) {
		fprintf(stderr, "Could not get dimension size for %s. "
			"NetCDF error message is:%s\n",
			name, nc_strerror(status));
		longjmp(error_env, NNCDF_ERROR);
	    }
	    sz *= l;
	}
	if ( !(uPtr = MALLOC(sz * sizeof(int))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
    }
    if ((status = nc_get_var_uchar(ncid, varid, uPtr)) != 0) {
	fprintf(stderr, "Could not get value for %s.  "
		"NetCDF error message is: %s\n", name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return uPtr;
}

/* Retrieve an integer variable from a NetCDF file. See nnetcdf (3). */
int * NNC_Get_Var_Int(int ncid, const char *name, int *iPtr, jmp_buf error_env)
{
    int varid;		/* Variable identifier */
    int status;

    if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !iPtr ) {
	int ndims;
	static int *dimidPtr;
	int *d, *de;
	size_t sz;

	if ((status = nc_inq_varndims(ncid, varid, &ndims)) != 0) {
	    fprintf(stderr, "Could not get dimension count for %s. "
		    "NetCDF error message is: %s\n",
		    name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ( !(dimidPtr = REALLOC(dimidPtr, ndims * sizeof(int))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ((status = nc_inq_vardimid(ncid, varid, dimidPtr)) != 0) {
	    fprintf(stderr, "Could not get dimensions for %s. "
		    "NetCDF error message is: %s\n", name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	for (sz = 1, d = dimidPtr, de = d + ndims; d < de; d++) {
	    size_t l;

	    if ((status = nc_inq_dimlen(ncid, *d, &l)) != 0) {
		fprintf(stderr, "Could not get dimension size for %s. "
			"NetCDF error message is:%s\n",
			name, nc_strerror(status));
		longjmp(error_env, NNCDF_ERROR);
	    }
	    sz *= l;
	}
	if ( !(iPtr = MALLOC(sz * sizeof(int))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
    }
    if ((status = nc_get_var_int(ncid, varid, iPtr)) != 0) {
	fprintf(stderr, "Could not get value for %s.  "
		"NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return iPtr;
}

/* Retrieve an unsigned integer variable from a NetCDF file. See nnetcdf (3). */
unsigned * NNC_Get_Var_UInt(int ncid, const char *name, unsigned *iPtr,
	jmp_buf error_env)
{
    int varid;		/* Variable identifier */
    int status;

    if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !iPtr ) {
	int ndims;
	static int *dimidPtr;
	int *d, *de;
	size_t sz;

	if ((status = nc_inq_varndims(ncid, varid, &ndims)) != 0) {
	    fprintf(stderr, "Could not get dimension count for %s. "
		    "NetCDF error message is: %s\n",
		    name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ( !(dimidPtr = REALLOC(dimidPtr, ndims * sizeof(int))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ((status = nc_inq_vardimid(ncid, varid, dimidPtr)) != 0) {
	    fprintf(stderr, "Could not get dimensions for %s. "
		    "NetCDF error message is: %s\n", name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	for (sz = 1, d = dimidPtr, de = d + ndims; d < de; d++) {
	    size_t l;

	    if ((status = nc_inq_dimlen(ncid, *d, &l)) != 0) {
		fprintf(stderr, "Could not get dimension size for %s. "
			"NetCDF error message is:%s\n",
			name, nc_strerror(status));
		longjmp(error_env, NNCDF_ERROR);
	    }
	    sz *= l;
	}
	if ( !(iPtr = MALLOC(sz * sizeof(int))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
    }
    if ((status = nc_get_var_uint(ncid, varid, iPtr)) != 0) {
	fprintf(stderr, "Could not get value for %s.  "
		"NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return iPtr;
}

/* Retrieve a float variable from a NetCDF file. See nnetcdf (3). */
float * NNC_Get_Var_Float(int ncid, const char *name, float *fPtr,
	jmp_buf error_env)
{
    int varid;		/* Variable identifier */
    int status;

    if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !fPtr ) {
	int ndims;
	static int *dimidPtr;
	int *d, *de;
	size_t sz;

	if ((status = nc_inq_varndims(ncid, varid, &ndims)) != 0) {
	    fprintf(stderr, "Could not get dimension count for %s. "
		    "NetCDF error message is: %s\n",
		    name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ( !(dimidPtr = REALLOC(dimidPtr, ndims * sizeof(int))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ((status = nc_inq_vardimid(ncid, varid, dimidPtr)) != 0) {
	    fprintf(stderr, "Could not get dimensions for %s. "
		    "NetCDF error message is: %s\n",
		    name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	for (sz = 1, d = dimidPtr, de = d + ndims; d < de; d++) {
	    size_t l;

	    if ((status = nc_inq_dimlen(ncid, *d, &l)) != 0) {
		fprintf(stderr, "Could not get dimension size for %s. "
			"NetCDF error message is:%s\n",
			name, nc_strerror(status));
		longjmp(error_env, NNCDF_ERROR);
	    }
	    sz *= l;
	}
	if ( !(fPtr = MALLOC(sz * sizeof(float))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
    }
    if ((status = nc_get_var_float(ncid, varid, fPtr)) != 0) {
	fprintf(stderr, "Could not get value for %s.  "
		"NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return fPtr;
}

/* Retrieve a double variable from a NetCDF file. See nnetcdf (3). */
double * NNC_Get_Var_Double(int ncid, const char *name, double *dPtr,
	jmp_buf error_env)
{
    int varid;		/* Variable identifier */
    int status;

    if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !dPtr ) {
	int ndims;
	static int *dimidPtr;
	int *d, *de;
	size_t sz;

	if ((status = nc_inq_varndims(ncid, varid, &ndims)) != 0) {
	    fprintf(stderr, "Could not get dimension count for %s. "
		    "NetCDF error message is: %s\n",
		    name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ( !(dimidPtr = REALLOC(dimidPtr, ndims * sizeof(int))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
	if ((status = nc_inq_vardimid(ncid, varid, dimidPtr)) != 0) {
	    fprintf(stderr, "Could not get dimensions for %s. "
		    "NetCDF error message is: %s\n",
		    name, nc_strerror(status));
	    longjmp(error_env, NNCDF_ERROR);
	}
	for (sz = 1, d = dimidPtr, de = d + ndims; d < de; d++) {
	    size_t l;

	    if ((status = nc_inq_dimlen(ncid, *d, &l)) != 0) {
		fprintf(stderr, "Could not get dimension size for %s."
			" NetCDF error message is:%s\n",
			name, nc_strerror(status));
		longjmp(error_env, NNCDF_ERROR);
	    }
	    sz *= l;
	}
	if ( !(dPtr = MALLOC(sz * sizeof(double))) ) {
	    fprintf(stderr, "Could not allocate dimension array for %s\n",
		    name);
	    longjmp(error_env, NNCDF_ERROR);
	}
    }
    if ((status = nc_get_var_double(ncid, varid, dPtr)) != 0) {
	fprintf(stderr, "Could not get value for %s. "
		" NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return dPtr;
}

/* Get a string attribute associated with a NetCDF variable. See nnetcdf (3). */
char * NNC_Get_Att_String(int ncid, const char *name, const char *att,
	jmp_buf error_env)
{
    int varid;
    int status;
    size_t len;
    char *val;

    if (strcmp(name, "NC_GLOBAL") == 0) {
	varid = NC_GLOBAL;
    } else if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_inq_attlen(ncid, varid, att, &len)) != 0) {
	fprintf(stderr, "Could not get string length for %s of %s."
		" NetCDF error message is: %s\n",
		att, name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !(val = MALLOC(len + 1)) ) {
	fprintf(stderr, "Allocation failed for %s\n",name);
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_get_att_text(ncid, varid, att, val)) != 0) {
	fprintf(stderr, "Could not get %s for %s."
		" NetCDF error message is: %s\n",
		att, name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    val[len] = '\0';
    return val;
}

/* Get integer attribute associated with a NetCDF variable. See nnetcdf (3). */
int *NNC_Get_Att_Int(int ncid, const char *name, const char *att,
	jmp_buf error_env)
{
    int varid;
    int status;
    size_t len;
    int *i;

    if (strcmp(name, "NC_GLOBAL") == 0) {
	varid = NC_GLOBAL;
    } else if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_inq_attlen(ncid, varid, att, &len)) != 0) {
	fprintf(stderr, "Could not get string length for %s of %s."
		" NetCDF error message is: %s\n",
		att, name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !(i = CALLOC(len, sizeof(int))) ) {
	fprintf(stderr, "Allocation failed for %s\n",name);
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_get_att_int(ncid, varid, att, i)) != 0) {
	fprintf(stderr, "Could not get %s attribute for %s."
		" NetCDF error message is: %s\n",
		att, name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return i;
}

/*
   Get unsigned integer attribute associated with a NetCDF variable.
   See nnetcdf (3).
 */

unsigned *NNC_Get_Att_UInt(int ncid, const char *name, const char *att,
	jmp_buf error_env)
{
    int varid;
    int status;
    size_t len;
    unsigned *i;

    if (strcmp(name, "NC_GLOBAL") == 0) {
	varid = NC_GLOBAL;
    } else if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_inq_attlen(ncid, varid, att, &len)) != 0) {
	fprintf(stderr, "Could not get attribute length for %s of %s."
		" NetCDF error message is: %s\n",
		att, name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !(i = CALLOC(len, sizeof(unsigned))) ) {
	fprintf(stderr, "Allocation failed for %s\n",name);
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_get_att_uint(ncid, varid, att, i)) != 0) {
	fprintf(stderr, "Could not get %s attribute for %s."
		" NetCDF error message is: %s\n",
		att, name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return i;
}

/* Get a float attribute associated with a NetCDF variable. See nnetcdf (3). */
float *NNC_Get_Att_Float(int ncid, const char *name, const char *att,
	jmp_buf error_env)
{
    int varid;
    int status;
    size_t len;
    float *v;

    if (strcmp(name, "NC_GLOBAL") == 0) {
	varid = NC_GLOBAL;
    } else if ((status = nc_inq_varid(ncid, name, &varid)) != 0) {
	fprintf(stderr, "No variable named %s. NetCDF error message is: %s\n",
		name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_inq_attlen(ncid, varid, att, &len)) != 0) {
	fprintf(stderr, "Could not get attribute length for %s of %s."
		" NetCDF error message is: %s\n",
		att, name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    if ( !(v = CALLOC(len, sizeof(float))) ) {
	fprintf(stderr, "Allocation failed for %s\n",name);
	longjmp(error_env, NNCDF_ERROR);
    }
    if ((status = nc_get_att_float(ncid, varid, att, v)) != 0) {
	fprintf(stderr, "Could not get %s attribute for %s."
		" NetCDF error message is: %s\n",
		att, name, nc_strerror(status));
	longjmp(error_env, NNCDF_ERROR);
    }
    return v;
}

